---
title: "Gene expression analysis, Greer 2021"
output: html_notebook
---
<style type="text/css">
.main-container {
  max-width: 1000px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{css}
code {white-space:pre !important; overflow-x:auto}
```

```{r libraries, message=FALSE, warning=FALSE}
library(magrittr)
library(DESeq2)
library(RColorBrewer)
library(gplots)
library(ggplot2)
library(rstudioapi)
library(dplyr)
library(stringr)
library(pheatmap)
library(data.table)
library(RRHO)
library(gridExtra)
setwd(dirname(getActiveDocumentContext()$path))
```

```{r custom functions}
# function to filter out genes that have less than 10 reads in at least 50% of samples 
filter_by_count <- function(data) {
  if(!is.matrix(data)) {stop('data is not a matrix')}
  names <- colnames(data)
  filtered <- apply(data, 1, function(x) {
    passed <- x[x >= 10]
    if (length(passed) > length(x)/2) { return(x)  }
    else {return(rep(NA, length(x)))}    
  }) %>% t() %>% .[complete.cases(.),] %>% set_colnames(names)
  return(filtered)
}

# wrapper for DESeq2 results
formatRes <- function(result) {
  result <- result[order(result$padj),]
  result <- as.data.frame(result)
  result$rank <- -sign(result$log2FoldChange)*log10(result$pvalue)
  result$geneName <- row.names(result)
  return(result)
}

# plot RRHO maps
rrhoMapPlot <- function(rrho, filename=NULL, plotTitle=NULL, scale=c(0,0), xlab=NULL, ylab=NULL, noFeatures=F, silent=F, blank=F) {
  jet.colors <- colorRampPalette(c("#00007F", "blue", "#007FFF", "cyan", "#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000"))
  plotData <- reshape2::melt(rrho$hypermat.by)
  p <- ggplot()
  if (blank) { p <- p + geom_raster(data=plotData, aes(x=Var1, y=Var2), fill = "grey75") }
  else       { p <- p + geom_raster(data=plotData, aes(x=Var1, y=Var2, fill=value)) }
  
  p <- p +
    coord_fixed(expand=FALSE) +
    theme_classic() +
    theme(legend.position = "none") +
    theme(axis.title=element_blank()) +
    theme(axis.text=element_blank()) +
    theme(axis.ticks=element_blank()) +
    theme(axis.line = element_blank()) +
    theme(plot.margin = margin(0, 0, 0, 0, "cm")) +
    labs(title=NULL)
  
  if (!is.null(plotTitle)) { p <- p + labs(title=plotTitle)}
  
  if(isTRUE(noFeatures)) { p <- p + labs(fill="")}
  else {
    if (!is.null(xlab)) { p <- p + theme(axis.title.x = element_text()) + labs(x=xlab)}
    if (!is.null(ylab)) { p <- p + theme(axis.title.y = element_text()) + labs(y=ylab)}
  }
  
  if(all(scale == c(0,0))) { p <- p + scale_fill_gradientn(colours = jet.colors(9))       }
  else { p <- p + scale_fill_gradientn(colours = jet.colors(9), limits=c(scale[1],scale[2]))  }
  
  if(!is.null(filename)) { ggsave(filename=filename, plot=p, device = 'png', width = 5, height = 5, dpi = 300, units = 'in')}
  if(silent) { p }  else { print(p)}
  
}
```

```{r load mRNA data}
sampleTable <- read.table(file = "./counts/SampleTable.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE) 
subtable <- sampleTable[sampleTable$type == "mrna" & sampleTable$study == 'greer',]
mrnaCounts <- lapply(1:nrow(subtable), function(x) {
                data <- read.table(file = paste0("./counts/",  subtable[x,"file"]), skip=4, row.names = 1) %>% select(1) %>% set_colnames(subtable[x,'abbr'])
                data$geneNames <- row.names(data) %>% str_replace("Gene:","")
                return(data)
})
mrnaCounts <- Reduce(function(df1, df2) {merge(df1, df2, by="geneNames", all=T)}, mrnaCounts)
row.names(mrnaCounts) <- mrnaCounts$geneNames
mrnaCounts$geneNames <- NULL
mrnaCounts <- as.matrix(mrnaCounts)
```

#### PCA plots for mRNAseq data  
```{r PCA plots, fig.align="center", fig.height=8, fig.width=8, out.width="50%"}
plotData <- filter_by_count(mrnaCounts) %>% rlog()
plotData <- plotData[order(rowVars(plotData), decreasing = TRUE), ] %>% .[1:1000,]
pca <- prcomp(t(plotData), scale. = TRUE)
percentVar <- pca$sdev^2/sum(pca$sdev^2)
PCAsummary <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], group = sampleTable[sampleTable$type == "mrna" & sampleTable$study == 'greer',"treatment"])

g <- ggplot(data = PCAsummary, aes_string(x = "PC1", y = "PC2", color = "group")) + 
  geom_point(size = 6) + 
  xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) +
  ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) +
  theme_bw() +
  theme(panel.grid = element_blank(), panel.border = element_rect(linetype = "solid", fill = NA, size = 1.5)) +
  theme(axis.ticks.length = unit(2, "mm"), axis.ticks = element_line(size = 1), axis.text = element_text(size = rel(1.25))) +
  theme(axis.title = element_text(size = rel(1.5)))+
  theme(legend.text = element_text(size = rel(1.2)), legend.title = element_blank())

ggsave(filename="./figures/PCA_mRNAseq.png", plot=g, device="png", units="in", dpi=300, width=8, height=8)
g

```
#### Differential Gene Expression Analysis (mRNAseq), ranked files
```{r DE analysis, warning=F, message=F}
DEresults <- list()

#-------------------------------------- Fed vs Starved -----------------------------------------------------------------
colData <- data.frame(sample = colnames(mrnaCounts)[1:12],
                      group = factor(c(rep("FE", 6), rep("ST", 6)), levels = c("FE","ST")),
                      stringsAsFactors = FALSE
)
countData <- filter_by_count(mrnaCounts[,1:12])
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design =~ group)
dds <- DESeq(dds)
#resultsNames(dds)
result <-  results(dds, name = 'group_ST_vs_FE', cooksCutoff = FALSE, independentFiltering = FALSE) %>% formatRes()
DEresults <- append(DEresults, list(STvsFE=result))
write.csv(result, file="mRNAseq_Starved_vs_Fed.csv")

#---------------------------------- Empty vector vs siRNA knockdowns -------------------------------------------------
colData <- data.frame(sample = colnames(mrnaCounts)[13:ncol(mrnaCounts)],
                      group = factor(c(rep("C27", 4), rep("EV", 4), rep("E02", 4)), levels = c("EV","E02","C27")),
                      stringsAsFactors = FALSE
)
countData <- filter_by_count(mrnaCounts[,13:ncol(mrnaCounts)])
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design =~ group)
dds <- DESeq(dds)
#resultsNames(dds)
result <-  results(dds, name = 'group_E02_vs_EV', cooksCutoff = FALSE, independentFiltering = FALSE) %>% formatRes()
DEresults <- append(DEresults, list(E02vsEV=result))
write.csv(result, file="mRNAseq_E02_vs_EV.csv")
result <-  results(dds, name = 'group_C27_vs_EV', cooksCutoff = FALSE, independentFiltering = FALSE) %>% formatRes()
DEresults <- append(DEresults, list(C27vsEV=result))
write.csv(result, file="mRNAseq_C27_vs_EV.csv")

#save rank files
junk <- lapply(names(DEresults), function(name) {
  rank <- DEresults[[name]][, c('geneName', 'rank')] %>% setNames(c('NAME','RANK'))
  rank <- rank[order(rank$RANK),]
  write.table(rank, file=paste0("./ranked_files/", name, "mRNA.rnk"), row.names = F, quote = F, sep = "\t") 
}) %>% rm()
```

#### RRHO maps, mRNAseq
```{r rrho maps}
# load data
rankFiles <- c("C27vsEVmRNA.rnk","E02vsEVmRNA.rnk", "STvsFEmRNA.rnk")
ranklist <- lapply(rankFiles, function(file) {
  data <- read.table(file = paste0("./ranked_files/", file), skip = 1, stringsAsFactors = FALSE) %>% set_colnames(c("geneName", str_sub(file, 0, -9)))
})
ranklist <- Reduce(function(df1, df2) {merge(df1, df2, by="geneName", all=T)}, ranklist)
ranklist <- ranklist[complete.cases(ranklist),]

rrhoMaps <- lapply(2:ncol(ranklist), function(i) {
  output <- lapply(2:ncol(ranklist), function(j)  {
    x.list <- ranklist[,c("geneName",colnames(ranklist)[i])]  
    y.list <- ranklist[,c("geneName",colnames(ranklist)[j])]
    map <- RRHO(list1=x.list, list2=y.list, stepsize = 50, plots=F, BY=T, log10.ind=T, alternative="two.sided")
  })  %>% set_names(colnames(ranklist)[2:ncol(ranklist)])
}) %>% set_names(colnames(ranklist)[2:ncol(ranklist)])

rrhoMap_scale <- c(0,0)
for(i in 1:length(rrhoMaps)){
  for(j in 1:length(rrhoMaps[[i]])){
    scale_max <- max(rrhoMaps[[i]][[j]]$hypermat.by[is.finite(rrhoMaps[[i]][[j]]$hypermat.by)], na.rm=T)
    scale_min <- min(rrhoMaps[[i]][[j]]$hypermat.by[is.finite(rrhoMaps[[i]][[j]]$hypermat.by)], na.rm=T)
    if(scale_max > rrhoMap_scale[2]) { rrhoMap_scale[2]=scale_max }
    if(scale_min < rrhoMap_scale[1]) { rrhoMap_scale[1]=scale_min }
  }
}

dopePlots <- lapply(1:length(rrhoMaps), function(i) {
  output <- lapply(1:length(rrhoMaps[[i]]), function(j){
    rrho <- rrhoMaps[[i]][[j]]
    x.name <- names(rrhoMaps)[i]
    y.name <- names(rrhoMaps[[i]])[j]
    if(i == j) {
      plot <- rrhoMapPlot(rrho=rrho, filename=paste0("./figures/RRHO_maps_mRNAseq/",x.name,"_vs_",y.name,".png"),
                          noFeatures=T, blank=T, silent=T)      
    }
    else {
      plot <- rrhoMapPlot(rrho=rrho, filename=paste0("./figures/RRHO_maps_mRNAseq/",x.name,"_vs_",y.name,".png"),
                          noFeatures=T, silent=T)    
    }
  }) %>% set_names(names(rrhoMaps[[i]]))
}) %>% set_names(names(rrhoMaps))

png(filename = "./figures/RRHO_maps_mRNAseq/rrho_matrix.png", width = 10, height = 10, units="in", res = 300)
grid.newpage()
panels <- arrangeGrob(grobs = unlist(dopePlots, recursive = F), ncol=3, padding=0, respect = T)
row_lab <- tableGrob(names(dopePlots), theme= ttheme_minimal(core = list(fg_params=list(rot=90))))
col_lab <- tableGrob(t(c("",names(dopePlots))), theme= ttheme_minimal())
combined <- gtable_cbind(row_lab, panels, size="last")
combined <- gtable_rbind(col_lab, combined, size="last")
grid.draw(combined)
invisible(dev.off())
```

Selected RRHO maps, color intensities are on the same scale  
```{r, fig.height=3, fig.width=6, dpi=300}
# selected RRHO
rrhoMap_scale <- c(0,0)
for(name in names(rrhoMaps$STvsFE)[1:2]){
  scale_max <- max(rrhoMaps$STvsFE[[name]]$hypermat.by[is.finite(rrhoMaps$STvsFE[[name]]$hypermat.by)], na.rm=T)
  scale_min <- min(rrhoMaps$STvsFE[[name]]$hypermat.by[is.finite(rrhoMaps$STvsFE[[name]]$hypermat.by)], na.rm=T)
  if(scale_max > rrhoMap_scale[2]) { rrhoMap_scale[2]=scale_max }
  if(scale_min < rrhoMap_scale[1]) { rrhoMap_scale[1]=scale_min }
}

plots <- lapply(names(rrhoMaps$STvsFE)[1:2], function(name) {
    rrho <- rrhoMaps$STvsFE[[name]]
    x.name <- "Starved     <=>       Fed"
    y.name <- name %>% strsplit("vs") %>% unlist() %>% paste0(collapse='    <=>    ')
    y.name <- strsplit(y.name, "vs")
    plot <- rrhoMapPlot(rrho=rrho, scale = rrhoMap_scale, noFeatures=F, silent=T, xlab=x.name, ylab= y.name)    
    
}) %>% set_names(names(rrhoMaps$STvsFE)[1:2])

png(filename = "./figures/RRHO_maps_mRNAseq/rrho_KDs_sameScale.png", width = 4, height = 2, units="in", res = 300)
grid.arrange(grobs=plots, ncol=2, nrow=1)
invisible(dev.off())

grid.arrange(grobs=plots, ncol=2, nrow=1)

```
Selected RRHO maps, each panel has its own scale of color intensities  
```{r, fig.height=3, fig.width=6, dpi=300}
plots <- lapply(names(rrhoMaps$STvsFE)[1:2], function(name) {
    rrho <- rrhoMaps$STvsFE[[name]]
    x.name <- "Starved     <=>       Fed"
    y.name <- name %>% strsplit("vs") %>% unlist() %>% paste0(collapse='    <=>    ')
    y.name <- strsplit(y.name, "vs")
    plot <- rrhoMapPlot(rrho=rrho, noFeatures=F, silent=T, xlab=x.name, ylab= y.name)    
    
}) %>% set_names(names(rrhoMaps$STvsFE)[1:2])

png(filename = "./figures/RRHO_maps_mRNAseq/rrho_KDs.png", width = 4, height = 2, units="in", res = 300)
grid.arrange(grobs=plots, ncol=2, nrow=1)
invisible(dev.off())

grid.arrange(grobs=plots, ncol=2, nrow=1)
```

#### Comparing with other studies
Sural et. al. 2019. HSF-1 overexpression  
```{r Sural2019, include=FALSE}
# loading mRNAseq count data
subtable <- sampleTable[sampleTable$type == "mrna" & sampleTable$study == 'sural2019' & sampleTable$treatment != 'HSF_OE',]
mrnaCounts_hsfOE <- lapply(1:nrow(subtable), function(x) {
                data <- read.table(file = paste0("./counts/",  subtable[x,"file"]), skip=4, row.names = 1) %>% select(1) %>% set_colnames(subtable[x,'abbr'])
                data$geneNames <- row.names(data) %>% str_replace("Gene:","")
                return(data)
})
mrnaCounts_hsfOE <- Reduce(function(df1, df2) {merge(df1, df2, by="geneNames", all=T)}, mrnaCounts_hsfOE)
row.names(mrnaCounts_hsfOE) <- mrnaCounts_hsfOE$geneNames
mrnaCounts_hsfOE$geneNames <- NULL
mrnaCounts_hsfOE <- as.matrix(mrnaCounts_hsfOE)

# Differential gene expression analysis
#---------------------------------- HSF-1 overexpression -------------------------------------------------
colData <- data.frame(sample = colnames(mrnaCounts_hsfOE),
                      group = factor(c(rep("overexp", 3), rep("control", 3)), levels = c("control","overexp")),
                      stringsAsFactors = FALSE
)
countData <- filter_by_count(mrnaCounts_hsfOE)
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design =~ group)
dds <- DESeq(dds)
result <-  results(dds, name = 'group_overexp_vs_control', cooksCutoff = FALSE, independentFiltering = FALSE) %>% formatRes()
write.csv(result, file="mRNAseq_hsf1OE_overexp_vs_control.csv")
rank <- result[, c('geneName', 'rank')] %>% setNames(c('NAME','RANK'))
rank <- rank[order(rank$RANK),]
write.table(rank, file=paste0("./ranked_files/", 'hsf_overexpr_vs_control', "_mRNA.rnk"), row.names = F, quote = F, sep = "\t") 

# combine ranked files from different studies together
ranklist <- lapply(list.files("./ranked_files"), function(file) {
  data <- read.table(file = paste0("./ranked_files/", file), skip = 1, stringsAsFactors = FALSE) %>% set_colnames(c("geneName", str_sub(file, 0, -9)))
})
ranklist <- Reduce(function(df1, df2) {merge(df1, df2, by="geneName", all=T)}, ranklist)
ranklist <- ranklist[complete.cases(ranklist),]
colnames(ranklist)[which(colnames(ranklist) == 'hsf_overexpr_vs_control_', arr.ind = T)] <- "HSFOEvsC"

rrhoMaps <- lapply(2:ncol(ranklist), function(i) {
  output <- lapply(2:ncol(ranklist), function(j)  {
    x.list <- ranklist[,c("geneName",colnames(ranklist)[i])]  
    y.list <- ranklist[,c("geneName",colnames(ranklist)[j])]
    #print(paste(colnames(ranklist)[i], colnames(ranklist)[j]))
    map <- RRHO(list1=x.list, list2=y.list, stepsize = 50, plots=F, BY=T, log10.ind=T, alternative="two.sided")
  })  %>% set_names(colnames(ranklist)[2:ncol(ranklist)])
}) %>% set_names(colnames(ranklist)[2:ncol(ranklist)])

dopePlots <- lapply(1:length(rrhoMaps), function(i) {
  output <- lapply(1:length(rrhoMaps[[i]]), function(j){
    rrho <- rrhoMaps[[i]][[j]]
    x.name <- names(rrhoMaps)[i]
    y.name <- names(rrhoMaps[[i]])[j]
    if(i == j) {
      plot <- rrhoMapPlot(rrho=rrho, noFeatures=T, blank=T, silent=T)      
    }
    else {
      plot <- rrhoMapPlot(rrho=rrho, noFeatures=T, silent=T)    
    }
  }) %>% set_names(names(rrhoMaps[[i]]))
}) %>% set_names(names(rrhoMaps))

png(filename = "./figures/RRHO_maps_mRNAseq/rrho_matrix_multistudy1.png", width = 10, height = 10, units="in", res = 300)
grid.newpage()
panels <- arrangeGrob(grobs = unlist(dopePlots, recursive = F), ncol=4, padding=0, respect = T)
row_lab <- tableGrob(names(dopePlots), theme= ttheme_minimal(core = list(fg_params=list(rot=90))))
col_lab <- tableGrob(t(c("",names(dopePlots))), theme= ttheme_minimal())
combined <- gtable_cbind(row_lab, panels, size="last")
combined <- gtable_rbind(col_lab, combined, size="last")
grid.draw(combined)
invisible(dev.off())

```

Selected RRHO maps, color intensities are on the same scale     
```{r, fig.height=3, fig.width=9, dpi=300}
rrhoMap_scale <- c(0,0)
for(name in names(rrhoMaps$HSFOEvsC)[c(1,2,4)]){
  scale_max <- max(rrhoMaps$HSFOEvsC[[name]]$hypermat.by[is.finite(rrhoMaps$HSFOEvsC[[name]]$hypermat.by)], na.rm=T)
  scale_min <- min(rrhoMaps$HSFOEvsC[[name]]$hypermat.by[is.finite(rrhoMaps$HSFOEvsC[[name]]$hypermat.by)], na.rm=T)
  if(scale_max > rrhoMap_scale[2]) { rrhoMap_scale[2]=scale_max }
  if(scale_min < rrhoMap_scale[1]) { rrhoMap_scale[1]=scale_min }
}

plots <- lapply(names(rrhoMaps$HSFOEvsC)[c(1,2,4)], function(name) {
    rrho <- rrhoMaps$HSFOEvsC[[name]]
    x.name <- "hsf OE     <=>       ctl"
    y.name <- name %>% strsplit("vs") %>% unlist() %>% paste0(collapse='    <=>    ')
    y.name <- strsplit(y.name, "vs")
    plot <- rrhoMapPlot(rrho=rrho, scale=rrhoMap_scale, noFeatures=F, silent=T, xlab=x.name, ylab= y.name)    
}) %>% set_names(names(rrhoMaps$HSFOEvsC)[c(1,2,4)])

png(filename = "./figures/RRHO_maps_mRNAseq/rrho_HSF_overexpression_sameScale.png", width = 6, height = 2, units="in", res = 300)
grid.arrange(grobs=plots, ncol=3, nrow=1)
invisible(dev.off())

grid.arrange(grobs=plots, ncol=3, nrow=1)

```

Selected RRHO maps, each panel has its own scale of color intensities  
```{r, fig.height=3, fig.width=9, dpi=300}
plots <- lapply(names(rrhoMaps$HSFOEvsC)[c(1,2,4)], function(name) {
    rrho <- rrhoMaps$HSFOEvsC[[name]]
    x.name <- "hsf OE     <=>       ctl"
    y.name <- name %>% strsplit("vs") %>% unlist() %>% paste0(collapse='    <=>    ')
    y.name <- strsplit(y.name, "vs")
    plot <- rrhoMapPlot(rrho=rrho, noFeatures=F, silent=T, xlab=x.name, ylab= y.name)    
    
}) %>% set_names(names(rrhoMaps$HSFOEvsC)[c(1,2,4)])

png(filename = "./figures/RRHO_maps_mRNAseq/rrho_HSF_overexpression.png", width = 6, height = 2, units="in", res = 300)
grid.arrange(grobs=plots, ncol=3, nrow=1)
invisible(dev.off())

grid.arrange(grobs=plots, ncol=3, nrow=1)

```

